core/src/main/scala/org/apache/spark/Dependency.scala

```diff
@@ -76,6 +78,15 @@ class ShuffleDependency[K, V, C](
 
   override def rdd: RDD[Product2[K, V]] = _rdd.asInstanceOf[RDD[Product2[K, V]]]
 
+  /**
+   * The key, value and combiner classes are serialized so that shuffle manager
+   * implementation can use the information to build
+   */
+  val keyClassName: String = reflect.classTag[K].runtimeClass.getName
+  val valueClassName: String = reflect.classTag[V].runtimeClass.getName
+  // Note: It's possible that the combiner class tag is null, if the combineByKey
+  // methods in PairRDDFunctions are used instead of combineByKeyWithClassTag.
+  val combinerClassName: Option[String] = Option(reflect.classTag[C]).map(_.runtimeClass.getName)
```

Comment:
If we make this `private[spark]`, that will prevent developers from creating pluggable `ShuffleManager` implementations, see `ShuffleManager.registerShuffle()`. The `ShuffleDependency` is annotated with `@DeveloperApi` which should signal to developers that these attributes can (and likely will) change over time, hopefully minimally.

Another approach would be to change the `registerShuffle` signature to be something like, e.g.

```
def registerShuffle[K, V, C](
      shuffleId: Int,
      numMaps: Int,
      dependency: ShuffleDependency[K, V, C],
      keyClassName: String,
      valueClassName: String,
      combinerClassName: Option[String]): ShuffleHandle
```

..instead of passing that information in the dependency.